Plug leaking function_records in cpp_function initialization in case of exceptions (found by Valgrind in #2746)

YannickJadoul opened the pull request:
## Description

Another leak fix taken out of #2746.

~~I am not entirely happy with this solution yet, mostly because we are not reusing `destruct(function_record *)`. I tried adding this as the `unique_ptr`'s deleter, but the issue is that the `char *`s only get copied along the way, and are not owned before those copies.~~ (kind of fixed that)
~~Also, `rec->data` can still leak as `rec->free_data` is not called after an exception.~~ (fixed that)

## Suggested changelog entry:

```rst
Resolved memory leak in cpp_function initialization when exceptions occured.
```

YannickJadoul commented on include/pybind11/pybind11.h:
```diff
@@ -126,7 +126,8 @@ class cpp_function : public function {
 
         /* Store the function including any extra state it might have (e.g. a lambda capture object) */
-        auto rec = make_function_record();
+        auto unique_rec = make_function_record();
```
Comment from #2746:

- @YannickJadoul (regarding `function_record::data`/`function_record::free_data`): This can still leak, I have just realized.

  It's very annoying that we cannot just have a `std::unique_ptr<detail::function_record, void (*)(detail::function_record *)>`, with destruct as deleter. The issue is that we cannot do this before all the `char *`s are `strdup`ed, and that 1) is only done in `initialize_generic`, and 2) the `repr` calls during `strdup` can also throw.

bstaletic replied:
Making `make_function_record()` return a `py::capsule` already wouldn't help, right?

YannickJadoul replied:
That would have the same issues as `std::unique_ptr<detail::function_record, void (*)(detail::function_record *)>`, I think? I.e., when do you know for sure that those strings have been `strdup`ed?

YannickJadoul replied:
I think I fixed it; trying to demonstrate these leaks in #2746's 8e8fc6a2cfd5e8e8ce92f4a07217435490e6649a

YannickJadoul commented:
So, this should be ready to review:

- 1d56f738dc085197184e2cb722a4f3f992257295 in #2746 demonstrates some extra leaks that weren't fixed yet
- 6c8e2f16da10517d5cb035e6034fa2b645699a7e shows that the last commit does fix it, now

The one question I now have, is whether this fix is worth it. As far as I know, all of these leaks are actually errors by the pybind11 user, and can't/shouldn't happen during actual use, if the library with pybind11 is correctly designed and programmed. So do we want to incur this overhead?
Then again, this overhead is only during `cpp_function` initialization, so it shouldn't be on any critical path, either. So I lean towards merging this PR. But open to alternative suggestions!

rwgk commented:
Hi @YannickJadoul,

Is this sentence in the PR top comment still true?
> Also, rec->data can still leak as rec->free_data is not called after an exception.

What do you think about the diff below, to make it easier for the compiler/optimizer to see that a runtime bool works here?

```
diff --git a/include/pybind11/pybind11.h b/include/pybind11/pybind11.h
index 114f72b..e57dccc 100644
--- a/include/pybind11/pybind11.h
+++ b/include/pybind11/pybind11.h
@@ -120,7 +120,7 @@ protected:
     PYBIND11_NOINLINE unique_function_record make_function_record() {
         // `destruct<true>(function_record)`: `initialize_generic` copies strings and
         // takes care of cleaning up in case of exceptions.
-        return unique_function_record(new detail::function_record(), destruct<false>);
+        return unique_function_record(new detail::function_record(), destruct_no_delete_strings);
     }
 
     /// Special internal constructor for functors, lambda functions, etc.
@@ -392,7 +392,7 @@ protected:
             rec->def->ml_flags = METH_VARARGS | METH_KEYWORDS;
 
             capsule rec_capsule(unique_rec.release(), [](void *ptr) {
-                destruct((detail::function_record *) ptr);
+                destruct_with_delete_strings((detail::function_record *) ptr);
             });
             guarded_strdup.release();
 
@@ -489,9 +489,16 @@ protected:
         }
     }
 
+    static void destruct_with_delete_strings(detail::function_record *rec) {
+        destruct_impl(rec, true);
+    }
+
+    static void destruct_no_delete_strings(detail::function_record *rec) {
+        destruct_impl(rec, false);
+    }
+
     /// When a cpp_function is GCed, release any memory allocated by pybind11
-    template <bool DeleteStrings = true>
-    static void destruct(detail::function_record *rec) {
+    static void destruct_impl(detail::function_record *rec, bool delete_strings) {
         // If on Python 3.9, check the interpreter "MICRO" (patch) version.
         // If this is running on 3.9.0, we have to work around a bug.
         #if !defined(PYPY_VERSION) && PY_MAJOR_VERSION == 3 && PY_MINOR_VERSION == 9
@@ -504,7 +511,7 @@ protected:
                 rec->free_data(rec);
             // During initialization, these strings might not have been copied yet,
             // so they cannot be freed. Once the function has been created, they can.
-            if (DeleteStrings) {
+            if (delete_strings) {
                 std::free((char *) rec->name);
                 std::free((char *) rec->doc);
                 std::free((char *) rec->signature);
```

YannickJadoul commented:
> Hi @YannickJadoul,
> 
> Is this sentence in the PR top comment still true?
> 
> > Also, rec->data can still leak as rec->free_data is not called after an exception.

Nope that's what I fixed, just today :-) Thanks, I fixed the original message!

> What do you think about the diff below, to make it easier for the compiler/optimizer to see that a runtime bool works here?

Hmm, not sure what you mean. In my version it is a compile-time bool, which should be easier to optimize, no?

That being said, your version should result in a smaller size overhead (at the cost of that extra runtime check), if we make sure that `destruct_impl` is not inlined (there's some pybind11 macro for that). So this is a valid trade-off to consider, that I hadn't yet thought of. Any thoughts?

rwgk commented:
> That being said, your version should result in a smaller size overhead (at the cost of that extra runtime check)

The overhead for the runtime check is almost certainly not measurable.
But the compiler or linker will have to work pretty hard to figure out how to avoid duplicating a sizable chunk of generated machine code.
Seems like a classical case of "template bloating" that's very easy to avoid, as suggested, and it makes the code clearer, too, at least for me.

YannickJadoul commented:
Alright, yes. Given other decisions in pybind11, prioritizing space over a tiny bit of performance makes sense to me! I've pulled this to runtime. (I have kept the default argument `= true`, though, indicating that that should be the default/normal case.)

YannickJadoul commented:
Alright, that commit passes Valgrind in #2746 (da8859d231f5ea78a6d10478ac12216fbf51be80) :-)

bstaletic commented on include/pybind11/pybind11.h:
```diff
@@ -114,9 +114,14 @@ class cpp_function : public function {
     object name() const { return attr("__name__"); }
 
 protected:
+    using unique_function_record = std::unique_ptr<detail::function_record, void (*)(detail::function_record *)>;
```
A function pointer has state. A non-capturing function object that just calls the deleter function has no state. The standard library has a way of not storing a stateless deleter which makes `sizeof(T*) == sizeof(std::unique_ptr<T>)`. Can we make the deleter stateless?

Reading a bit more into this, I guess storing the function pointer was just convenient, because lambdas in non-evaluated context are a C++20 thing. We'll have to fall back to C++98 function objects.

YannickJadoul replied:
Done, I think.

bstaletic commented on include/pybind11/pybind11.h:
```diff
@@ -207,7 +214,8 @@ class cpp_function : public function {
 
         /* Register the function with Python from generic (non-templated) code */
-        initialize_generic(rec, signature.text, types.data(), sizeof...(Args));
+        // Pass on the ownership over the `unique_rec` to `initialize_generic`. `rec` stays valid.
```
I'm not too happy about the sort-of-implicit contract of "you take ownership, but can't yet destroy the object". I understand why this is the case, it's just not what "pass on the ownership" really means.

YannickJadoul replied:
No, I agree, but 1) it's already what was happening before (it's just syntactically explicit now), and 2) I wouldn't immediately know what to do otherwise? Return the pointer from `initialize_generic`? That also feels reasonably ugly/non-intuitive.

bstaletic replied:
Right. I was just thinking out loud.

bstaletic commented on include/pybind11/pybind11.h:
```diff
@@ -223,20 +231,48 @@ class cpp_function : public function {
+        }
+        char *operator()(const char *s) {
+            auto t = strdup(s);
+            strings.emplace_back(t);
```
Any reason to use `emplace_back` instead of `push_back` here?

YannickJadoul replied:
Nope (except for maybe `clang-tidy`'s `modernize-use-emplace`?) Any reason for `push_back`?

bstaletic replied:
It's a simpler mechanism when you already have a `T` and it also doesn't need to instantiate a variadic member function template. On the other hand `emplace()` can't provide its main advantage of constructing in place, because you're already passing a `T`.

Not a big deal either way. Some people prefer to use `push_back(t)` while others prefer the simpler rule of "always use `emplace`".

YannickJadoul replied:
Yeah, OK. I do agree, actually, if it's a pointer, but something tells me clang-tidy bit me, last time I tried it...

@henryiii, any preferences? You're the one that set up clang-tidy :-P

bstaletic replied:
I think clang-tidy would be fine with `push_back()` here, though I'm not really against leaving it as-is.

henryiii replied:
If clang-tidy wants emplace_back, then that's fine with me, though I think it should be happy with push_back for a pointer like this.

YannickJadoul replied:
Let me try. I don't think I actually tried this out.

YannickJadoul replied:
OK, you're right; clang-tidy passes! I must have misremembered/misinterpreted an old clang-tidy failure. This is better, indeed :-)

bstaletic commented on include/pybind11/pybind11.h:
```diff
@@ -223,20 +231,48 @@ class cpp_function : public function {
+
     /// Register a function call with Python (generic non-templated code goes here)
-    void initialize_generic(detail::function_record *rec, const char *text,
+    void initialize_generic(unique_function_record &&unique_rec, const char *text,
```
Why are you taking `std::unique_ptr` by reference instead of by-value? Remember that r-value references mean "maybe move".

Considering that we don't really want the unique_ptr here, rather just the pointer and a way to tell the unique_ptr it isn't in charge, maybe take by const-ref? A const-ref-unique-pointer has weird semantic meaning, transferring ownership from unique pointer to a capsule is probably equally weird.

YannickJadoul replied:
> Considering that we don't really want the unique_ptr here, rather just the pointer and a way to tell the unique_ptr it isn't in charge, maybe take by const-ref?

That not going to work, because it needs to give up ownership to the `capsule`, later. I really want to pass ownership, not just `initialize_generic` use it.

> Remember that r-value references mean "maybe move".

True. So, ordinarily by value?

YannickJadoul replied:
Combined with https://github.com/pybind/pybind11/pull/2756#discussion_r551519918, this is even worse, though, the accept by value :-/

At least with `unique_function_record &&`, you have the contract "we only move in case of success, and otherwise the caller keeps ownership"

bstaletic replied:
> > Considering that we don't really want the unique_ptr here, rather just the pointer and a way to tell the unique_ptr it isn't in charge, maybe take by const-ref?
> 
> That not going to work, because it needs to give up ownership to the `capsule`, later. I really want to pass ownership, not just `initialize_generic` use it.
> 

Right, but passing by mutable reference would work. Though you know what I think about mutable references.

> > Remember that r-value references mean "maybe move".
> 
> True. So, ordinarily by value?

I guess so.

YannickJadoul replied:
> Right, but passing by mutable reference would work. Though you know what I think about mutable references.

Is that any cleaner? At least now we know that we're "maybe moving"?

If you have a nice combo with https://github.com/pybind/pybind11/pull/2756#discussion_r551519918, I'll implemented it (or feel free to push something), but otherwise...

bstaletic replied:
> a nice combo with #2756 (comment)

Unfortunately not. Let's just go with pass by value.

YannickJadoul replied:
Added a comment on why we should probably not do so: https://github.com/pybind/pybind11/pull/2756/commits/9942614495cada974791019b1644d9bb07d21a8d

bstaletic commented on include/pybind11/pybind11.h:
```diff
@@ -223,20 +231,48 @@ class cpp_function : public function {
+            return t;
+        }
+        void release() {
+            strings.clear();
```
`strings.clear()` doesn't call `free()` on the cleared pointers. Should it?

YannickJadoul replied:
Nope, or it wouldn't be releasing! (This name is meant to mirror `unique_ptr::release()`.)

YannickJadoul commented:
@henryiii, @rwgk, @EricCousineau-TRI, can I gently ping? This is the last thing standing between us, and rebasing #2746 and soon having Valgrind checks run on all PRs :-)

rwgk commented:
Hi @YannickJadoul, a few days ago I offered to run this PR through Google's global testing system, but you discouraged it. I'd happily approve this PR after convincing myself that the tests come back clean.

Fully absorbing all the nuances of this very hairy change would cost me more time than I can give it. I trust that you and @bstaletic have done a great job. Given that we have comprehensive testing including sanitizers in place, getting more heads to look at the details doesn't seem like a productive division of labor.

Please let me know when it this is a good time to globally test this PR again.

YannickJadoul commented:
> Hi @YannickJadoul, a few days ago I offered to run this PR through Google's global testing system, but you discouraged it. I'd happily approve this PR after convincing myself that the tests come back clean.

Well, you asked that for #2746, and I informed you that this code was already tested (except for the last few additional commits that were not made because tests fail, but because reviewing the code found a logical corner case for leaks that wasn't covered by tests; not at Google either, since you said ASAN/LSAN came back clean?). Also, #2746 isn't supposed to merge this functionality; I'll rebase it on top of this, and #2746 will be a bare-bones change to CI and CMake.

It's the central part of our code, so the successful route is tested by all our tests. I don't believe more tests would tell us much more, since the nature of this PR is fixing leaks in exceptional cases (most often programming mistakes, actually). Though I'm of course not stopping you to run tests! (They're just not the right tool to fully assess this PR, I think.)

I was mainly asking for one or two more reviews and approval, so we can merge this and get full Valgrind runs in CI (like @bstaletic's, but he was involved in creating this, so I guess that only counts as half an approval?). I'm sure that before 2.6.2 is released, you'll still run Google's tests anyway?

rwgk commented:
> so I guess that only counts as half an approval?).

I look at it as a full approval: you still have two heads looking at the same thing.
Of course, more is always better, but also more expensive: diminishing returns.

> I'm sure that before 2.6.2 is released, you'll still run Google's tests anyway?

Running tests for one thing at a time has a lot of value: easier to find root causes.
My company is spending a lot of resources to make retesting easy, for humans. Let the machines to the work! That frees up human energy.

YannickJadoul commented:
Also: what the hell is going on with our CI/GitHub Actions?

```
/home/runner/work/_temp/9fcba6cd-dc0e-4f5e-a169-2da04a13086d.sh: line 1: /opt/hostedtoolcache/cmake/3.19.3/x64/cmake-3.19.3-Linux-aarch64/bin/cmake: cannot execute binary file: Exec format error
```

We're not running `aarch64`, you silly thing.

YannickJadoul commented:
> Running tests for one thing at a time has a lot of value: easier to find root causes.
> My company is spending a lot of resources to make retesting easy, for humans. Let the machines to the work! That frees up human energy.

That's a good argument, yes. But as argued, in this case, I'm not sure the machines can tell us anything more than we already know (unless there's specialized tests that try to stress test pybind11's exceptions?), and won't say much more interesting :-/

YannickJadoul commented:
Anyway, go head, @rwgk. But given that you already tested the logic (all that was added since is an extra cleanup), I'm not expecting anything there. But if it convinces you we can merge this, then why not.

rwgk commented:
> Also, #2746 isn't supposed to merge this functionality; I'll rebase it on top of this, and #2746 will be a bare-bones change to CI and CMake.

Thanks, good to know. I'll test this PR specifically. Our global testing doesn't use CMake at all.

YannickJadoul commented:
> Also: what the hell is going on with our CI/GitHub Actions?
> 
> ```
> /home/runner/work/_temp/9fcba6cd-dc0e-4f5e-a169-2da04a13086d.sh: line 1: /opt/hostedtoolcache/cmake/3.19.3/x64/cmake-3.19.3-Linux-aarch64/bin/cmake: cannot execute binary file: Exec format error
> ```
> 
> We're not running `aarch64`, you silly thing.

OK, fixed by rebasing onto #2790. The remaining failures are still #2774, so unrelated.

rwgk commented:
Quick confirmation: this PR ~by itself*** on top of `master` makes the tests under pybind11/tests run ASAN & MSAN clean. The only sanitizer errors I'm still seeing are from TSAN, as reported under #2754.

*** I'm also using #2409 and two small local patches I've been carrying for a while.

YannickJadoul commented:
> Quick confirmation: this PR ~by itself*** on top of `master` makes the tests under pybind11/tests run ASAN & MSAN clean.

Huh, that's funny. So it's not flagging the one issue we couldn't really track down in #2746, and had to work around? (https://github.com/pybind/pybind11/pull/2746/commits/c1831201c047bce4c72ff469c2a9b81a5b9f41f8) Maybe/hopefully it's a Valgrind fluke, then, but it was quite consistent, so not sure I'm believing that :-/

henryiii reviewed (approved):
I'm happy to get this in.

rwgk commented:
> Huh, that's funny. So it's not flagging the one issue we couldn't really track down in #2746, and had to work around? ([c183120](https://github.com/pybind/pybind11/commit/c1831201c047bce4c72ff469c2a9b81a5b9f41f8))

Nope. I ran ASAN with pretty strict options, but MSAN only with defaults. I could play more with the options ... later! Let's get this in as-is asap IMO, pending only the results of the global testing run.

YannickJadoul commented:
Thanks, @henryiii :-)

> Nope. I ran ASAN with pretty strict options, but MSAN only with defaults. I could play more with the options ... later!

No worries. It's the one thing @bstaletic and I couldn't figure out. So the plan is to get Valgrind in, with the workaround, then undo the workaround in another PR that can serve as bug report. So I propose we keep discussion of this one thing for that one undoing-the-workaround PR?

rwgk commented:
The global testing came back clean. @yannick, please merge!

This PR was already extremely useful for my work on #2672, to sanitize my new code (I found a couple bugs already).

rwgk commented:
FYI: After rolling out this PR Google-internal, our testing system discovered this (very minor) leak:
https://github.com/tensorflow/tensorflow/blob/10c06df9492c44062f3a6b0c04bec325c7738426/tensorflow/python/eager/pywrap_tfe_src.cc#L4335
I don't know how this PR unmasked that leak or how the test even succeeded before this PR. I mailed a fix. Let's see if the review sheds some light on this.

YannickJadoul commented:
Interesting; if you do find out anything relevant or need to link back to these changes, I'm very interested.
